tun/tun_illumos.go

```diff
@@ -0,0 +1,271 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright 2019 Joyent, Inc.
+ */
+
+// +build illumos,amd64
+
+package tun
+
+import (
+	"fmt"
+
+	"golang.org/x/sys/unix"
+	"os"
+)
+
+// NativeTun is the OS-specific implementation of a tun interface.
+type NativeTun struct {
+	tunFile *os.File /* TUN device file */
+	ip_fd   int      /* IP device fd */
+	name    string   /* Interface name */
+	mtu     int
+
+	events chan Event
+	errors chan error
+}
+
+// Name implements the Device interface
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+// Flush implements the Device interface
+func (tun *NativeTun) Flush() error {
+	// TODO: can flushing be implemented by buffering and using sendmmsg?
+	return nil
+}
+
+// Read implements the Device interface
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+	default:
+		_, read, _, err := unix.Getmsg(int(tun.tunFile.Fd()), nil, buf[offset:])
```

Comment:
Would you be better off with caching the SysConn and using the safer closures it provides, instead of using Fd(), which is potentially racy?